### PR TITLE
Learning

### PR DESCRIPTION
**Pull request by oninbo**
So many things were done

### CODE REVIEW

File: texting_ai.py [hunk 1]
```diff
@@ -163,23 +140,36 @@ def __init__(self, save_file_name: str):
             self.knowledge_base: Dict[str, Dict[str, List[str]]] = dict()
             json_manager.write(self.knowledge_base, save_file_name)
 
-    def sentence_to_pattern(self, sentence: str) -> str:
+    def _sentence_to_pattern(self, sentence: str) -> Optional[str]:
```
Comment (UnoYakshi) [inline texting_ai.py]: Consider renaming the function. Functions should express actions, like `_make_pattern_from_sentence()`.

File: texting_ai.py [hunk 2]
```diff
@@ -163,23 +140,36 @@ def __init__(self, save_file_name: str):
...
+
+        # is sentence simple?
+        if not self._is_simple(tagged):
+            return None
+
         for word, tag in tagged:
-            # S is for nouns and V is for verbs
-            if tag in ['S', 'V']:
-                nouns_and_verbs.append(text_processing.stem(word))
+            if tag in [self._parts_of_speech['noun'], self._parts_of_speech['verb'],
```
Comment (UnoYakshi) [inline texting_ai.py]: You might want to have a variable for this list instead of writing it in the `if` block.

File: texting_ai.py [hunk 3]
```diff
@@ -328,6 +291,49 @@ def __init__(self, *args: [LearningAgent, NounsFindingAgent]):
...
+        self._agent_callers: Dict[Type, 'function'] = dict()
+        self._agent_callers[NounsFindingAgent] = lambda **kwargs: \
+            kwargs.get('agent', None).get_replies(*self._agent_adapters[NounsFindingAgent](**kwargs))
+        self._agent_callers[LearningAgent] = lambda **kwargs: \
+            kwargs.get('agent', None).get_replies(*self._agent_adapters[LearningAgent](**kwargs))
+        self._agent_callers[RandomReplyAgent] = lambda **kwargs: \
+            kwargs.get('agent', None).get_reply(*self._agent_adapters[RandomReplyAgent](**kwargs))
+        self._agent_callers[RatingLearningAgent] = lambda **kwargs:\
+            kwargs.get('agent', None).get_rated_replies(*self._agent_adapters[RatingLearningAgent](**kwargs))
+        self._agent_callers[RatingRandomReplyAgent] = lambda **kwargs: \
```
Comment (UnoYakshi) [inline texting_ai.py]: You might want to refactor all these repeating lines one day soon...

File: texting_ai.py [hunk 4]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+                new_knowledge_base[pattern]: Dict[str, int] = dict()
+            for reply in rules.get('replies', []):
+                new_knowledge_base[pattern][reply] = init_good_reply_val
+            for reply in rules.get('black list', []):
+                new_knowledge_base[pattern][reply] = init_bad_reply_val
+        self.knowledge_base = new_knowledge_base
+        self.save_file_name = path_to_base_file
+        json_manager.write(self.knowledge_base, path_to_base_file)
+
+    def __init__(self, save_file_name: str, predecessor_save_file: str = ""):
```
Comment (UnoYakshi) [inline texting_ai.py]: Perhaps, a constructor should be the first function in a class.

File: texting_ai.py [hunk 5]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+
+    def get_rated_replies(self, input_text: str) -> Tuple[Dict[str, int]]:
+        """
+        Gets rated replies on given input text
+        :param input_text: text message from user
+        :return: replies and corresponding rating
+        """
+        result = dict()
+        all_patterns = list(self.knowledge_base.keys())
+        found_patterns = list(filter(lambda pattern: re.search(pattern, input_text, re.I),
```
Comment (UnoYakshi) [inline texting_ai.py]: Shouldn't it be one line?..

File: texting_ai.py [hunk 6]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+                                     all_patterns))
+        for found_pattern in found_patterns:
+            for reply, rating in self.knowledge_base[found_pattern].items():
+                result[reply] = rating
+
+        return result,
```
Comment (UnoYakshi) [inline texting_ai.py]: Looks weird.

File: texting_ai.py [hunk 7]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        self._max_weight = 2
+        # for multiplying weight of a given reply
+        self.__given_reply_multiplier = 2
+        self.__random_reply_divisor = 2
+        self._phrases_weights: Dict[str, int] = dict()
+        for phrase in self._all_phrases:
+            self._phrases_weights[phrase] = self._max_weight
+
+    def _decrease_weight(self, reply):
+        # decreasing weight of a chosen reply
```
Comment (UnoYakshi) [inline texting_ai.py]: It should be a docstring with annotations, not just a comment.

File: texting_ai.py [hunk 8]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+                k = 1
+            else:
+                k = math.floor(len(replies) / 2)
+
+            # adding a random number of additional phrases
+            # depending on no_empty_reply parameter
+            possible_replies = replies + random.choices(list(filter(
+                lambda x: x not in replies,
+                self._all_phrases)),
+                k=k)
```
Comment (UnoYakshi) [inline texting_ai.py]: Please, refactor it. It looks kind of unreadable.

File: texting_ai.py [hunk 9]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+
+    def get_rated_reply(self, rated_replies: Dict[str, int],
+                        replies: List[str], black_list: List[str],
+                        no_empty_reply: bool) -> Tuple[Optional[str]]:
+        """
+        Gets random reply from given rated and regular replies and all phrases
+        :param rated_replies: replies with rating
+        :param replies: replies without rating
+        :param black_list: replies that should not be chosen
+        :param no_empty_reply: flag that is True when there must be non-empty reply
```
Comment (UnoYakshi) [inline texting_ai.py]: Consider rewriting it (and ALL the similar stuff) like `either to allow an empty reply or not`.

File: texting_ai.py [hunk 10]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        :return: reply on None if it's not possible to get a reply
+        """
+        possible_replies: List[str] = list()
+        if rated_replies or replies:
+            random_replies = list(filter(lambda x:
+                                         not (replies and x in replies
+                                              or rated_replies and x in rated_replies),
+                                         self._all_phrases))
```
Comment (UnoYakshi) [inline texting_ai.py]: Please, refactor it.

File: texting_ai.py [hunk 11]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+            random_replies.sort(key=lambda x: self._phrases_weights[x],
+                                                                  reverse=True)
```
Comment (UnoYakshi) [inline texting_ai.py]: Again, there is no sense in separating the lines here.

File: texting_ai.py [hunk 12]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+            possible_replies = \
+             list(set(filter(lambda x: x not in black_list, replies
+                             + list(rated_replies.keys())
+                             # adding one phrase with the most weight
+                             + [random_replies[0]])))
```
Comment (UnoYakshi) [inline texting_ai.py]: Don't separate one-block logic parts like that.

File: texting_ai.py [hunk 13]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        elif no_empty_reply:
+            possible_replies = list(filter(lambda x: x not in black_list, self._all_phrases))
+
+        if possible_replies:
+            reply = random.choices(possible_replies, list(map(lambda x: rated_replies.get(x, 0)
+                                                              + self._phrases_weights.get(x, 0)
+                                                              if rated_replies.get(x, 0)
+                                                              + self._phrases_weights.get(x, 0) > 0 else 0,
+                                                              possible_replies)))[0]
```
Comment (UnoYakshi) [inline texting_ai.py]: Please, refactor it. You may consider extracting some values to variables.

File: texting_ai.py [hunk 14]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+
+class MessagesCounter:
+    """For control of messages frequency of the bot"""
+
+    # minimum number of messages between bot's replies
+    messages_period = 200
+    # number of all messages that bot received
+    messages_num = 0
+
+    def count_and_check(self) -> bool:
```
Comment (UnoYakshi) [inline texting_ai.py]: Check what?

File: texting_ai.py [hunk 15]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+    """
+    Checks if the text contains the calling construction
+    """
+    def __init__(self):
+        self.names = [
+            'рей',
+            'аянами',
+            'рей аянами',
+            'аянами рей'
+        ]
```
Comment (UnoYakshi) [inline texting_ai.py]: Firstly, this list can be `frozenlist()`. Second, it could be a constant somewhere.

File: texting_ai.py [hunk 16]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+
+    def check(self, text) -> bool:
```
Comment (UnoYakshi) [inline texting_ai.py]: Please, fix the annotation for `text`.

File: texting_ai.py [hunk 17]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+class ConversationController:
+    """
+    Controls how bot should reply on a given message depending on its source
+    """
+
+    @staticmethod
+    def _is_question(text) -> bool:
+        return re.search(r'\?', text)
+
+    def __init__(self, agent_pipeline: AgentPipeline):
```
Comment (UnoYakshi) [inline texting_ai.py]: Again, consider moving a constructor to the top of the class.

File: texting_ai.py [hunk 18]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        self._messages_counter = MessagesCounter()
+        self._call_checker = TextCallChecker()
+
+        self._agent_pipeline = agent_pipeline
+
+    def proceed_input_message(self, input_text: str,
+                              is_private: bool = False,
+                              is_call: bool = False) -> Optional[str]:
+        """
+        data flow when the message is received
```
Comment (UnoYakshi) [inline texting_ai.py]: The description barely describes what will happen.

File: texting_ai.py [hunk 19]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        :param input_text: text of the message
+        :param is_private: is message private?
+        :param is_call: does message contains calling construction?
+        :return: reply on message or None
+        """
+        is_call = is_call or self._call_checker.check(input_text)
+        no_empty_reply = \
+            True if self._is_question(input_text) or is_call else False
```
Comment (UnoYakshi) [inline texting_ai.py]: Why'd you make a newline here? I think it fits 120-200 characters.

File: texting_ai.py [hunk 20]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+        if is_call or is_private or self._messages_counter.count_and_check():
+            reply = self._agent_pipeline.get_reply(input_text, no_empty_reply=no_empty_reply)
+            if reply:
+                self._messages_counter.reset()
+        else:
+            reply = None
```
Comment (UnoYakshi) [inline texting_ai.py]: It can be `return None`.

File: texting_ai.py [hunk 21]
```diff
@@ -341,24 +347,331 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False) -> Optional[s
...
+RANDOM_REPLY_AGENT = RatingRandomReplyAgent(os.path.join(AGENT_LANGUAGE_PATH, 'sentences.json'))
+NOUNS_FINDING_AGENT = \
+            NounsFindingAgent(os.path.join(AGENT_LANGUAGE_PATH, 'sentences.json'),
+                              os.path.join(AGENT_LANGUAGE_PATH, 'nouns.json'))
+LEARNING_AGENT = RatingLearningAgent(os.path.join('data', 'rated_learning_model.json'),
+                                     os.path.join('data',
+                                                  'learning_model.json'))
+AGENTS_PIPELINE = \
+            AgentPipeline(LEARNING_AGENT, NOUNS_FINDING_AGENT, RANDOM_REPLY_AGENT)
+CONVERSATION_CONTROLLER = ConversationController(AGENTS_PIPELINE)
```
Comment (UnoYakshi) [inline texting_ai.py]: Consider moving the constant either to the top of the module (just right after module-docstring and imports) or even to a separate module.

File: texting_ai.py [hunk 22]
```diff
@@ -63,33 +61,19 @@ def __init__(self, phrases_json_path: str, nouns_json_path: str):
...
-
-    def get_reply(self, input_text: str, no_empty_reply: bool = False,
-                  black_list: Optional[List[str]] = None) -> Optional[str]:
+    def get_replies(self, input_text: str,
+                    black_list: Optional[List[str]] = None) -> Tuple[List[str]]:
         """
-        Returns text output by
+        Returns possible text outputs by
         search known nouns in the input text
         and giving predefined phrases as a reply
```
[the remark this answers is not in the document]
Comment (oninbo) [reply texting_ai.py]: What do you mean by separating the text?

File: bot.py [hunk 23]
```diff
@@ -104,20 +106,13 @@ def text_reply(message: telebot.types.Message) -> None:
...
-    # if private message
-    if message.chat.type == PRIVATE_MESSAGE:
-        is_reply = False
-    # TODO add forward handling
-    # if group message and not a reply on bot's message
-    elif not check_reply(BOT.get_me().id, message):
-        no_empty_reply = False
-
-    reply = texting_ai.PIPELINE.get_reply(text, no_empty_reply=no_empty_reply)
+    is_private = message.chat.type == PRIVATE_MESSAGE
```
[the remark this answers is not in the document]
Comment (oninbo) [reply bot.py]: Is that a compliment?

File: texting_ai.py [hunk 24]
```diff
@@ -63,33 +61,19 @@ def __init__(self, phrases_json_path: str, nouns_json_path: str):
...
-        self.last_used_reply = None
-
-        # for decreasing frequency of messages sent by bot
-        # TODO remove that to bot module as "typing"
-        self.message_counter = self.MESSAGES_PERIOD
-
-    def get_reply(self, input_text: str, no_empty_reply: bool = False,
-                  black_list: Optional[List[str]] = None) -> Optional[str]:
+    def get_replies(self, input_text: str,
+                    black_list: Optional[List[str]] = None) -> Tuple[List[str]]:
```
[the remark this answers is not in the document]
Comment (oninbo) [reply texting_ai.py]: Because of architecture

File: texting_ai.py [hunk 25]
```diff
@@ -105,55 +89,48 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False,
...
-        if reply_variants:
-            if no_empty_reply and len(reply_variants) == 1:
-                result_reply = reply_variants[0]
-            else:
-                while True:
-                    result_reply = random.choice(reply_variants)
-                    reply_variants.remove(result_reply)
+            for reply in black_list:
+                if reply in reply_variants:
+                    reply_variants = list(filter(lambda x: x != reply, reply_variants))
```
[the remark this answers is not in the document]
Comment (oninbo) [reply texting_ai.py]: It just removes all instances of `reply` from `reply_variants`

File: texting_ai.py [hunk 26]
```diff
@@ -105,55 +89,48 @@ def get_reply(self, input_text: str, no_empty_reply: bool = False,
...
+                        tagged_words))
+        if len(punctuation_symbols) > 1 or len(punctuation_symbols) == 1 \
+                and tagged_words[-1] != punctuation_symbols[0]:
+            return False
+
+        # are there any connecting words (союзы)?
+        connecting_words = \
+            list(filter(lambda tagged_word: tagged_word[1]
+                        == self._parts_of_speech['connecting words'],
+                        tagged_words))
```
[the remark this answers is not in the document]
Comment (oninbo) [reply texting_ai.py]: pylint sees

File: texting_ai.py (hunk 2 above)
Comment (oninbo) [reply texting_ai.py]: You mean putting `[self._parts_of_speech['noun'], self._parts_of_speech['verb']` inside a variable?

File: texting_ai.py (hunk 3 above)
Comment (oninbo) [reply texting_ai.py]: Oh yeah...

File: texting_ai.py (hunk 5 above)
Comment (oninbo) [reply texting_ai.py]: Why not?

File: texting_ai.py (hunk 6 above)
Comment (oninbo) [reply texting_ai.py]: It's a tuple

File: texting_ai.py (hunk 19 above)
Comment (oninbo) [reply texting_ai.py]: Pylint disagrees

File: texting_ai.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (oninbo) [reply texting_ai.py]: Well, it is still possible to write it in more optimal way

File: texting_ai.py (hunk 2 above)
Comment (UnoYakshi) [reply texting_ai.py]: Well, yes.

File: texting_ai.py (hunk 5 above)
Comment (UnoYakshi) [reply texting_ai.py]: It doesn't seem to be overcomplicated used in just a one line.

**Pull request merged**